helper/profiles/profiles.go

```diff
@@ -388,25 +473,106 @@ func (p *ProfileEngine) evaluateTypedField(ctx context.Context, history *Evaluat
 	}
 
 	sourceEval := sourceBuilder(ctx, p, obj)
+
 	defer sourceEval.Close(ctx)
 
-	// XXX (ascheel) - We will need to validate that history aligns with our
-	// expectations at this point in time, however, our actual builder
-	// initialization and validation call will be moved earlier to profile
-	// validation.
-	_, _, err := sourceEval.Validate(ctx)
+	accessedRequests, accessedResponses, err := sourceEval.Validate(ctx)
 	if err != nil {
 		return nil, fmt.Errorf("failed to validate source '%v': %w", source, err)
 	}
 
+	if len(accessedRequests) == 0 && len(accessedResponses) == 0 {
+		return nil, fmt.Errorf("no valid requests or responses found")
+	}
+
+	for _, req := range accessedRequests {
+		if req == "" {
+			return nil, fmt.Errorf("invalid empty request name found")
+		}
+	}
+
+	for _, resp := range accessedResponses {
+		if resp == "" {
+			return nil, fmt.Errorf("invalid empty response name found")
+		}
+	}
+
 	val, err := sourceEval.Evaluate(ctx, history)
 	if err != nil {
 		return nil, fmt.Errorf("failed to evaluate source '%v': %w", source, err)
 	}
 
-	// XXX (ascheel) - add handling of objType here; make sure val is
-	// assertable to the desired type, which might require another call to
-	// mapstructure here.
+	for outerName, innerMap := range history.Requests {
+		for reqName := range innerMap {
+			if _, ok := history.Responses[outerName][reqName]; !ok {
+				return nil, fmt.Errorf(
+					"history inconsistency: no response recorded for request '%s' in block '%s'",
+					reqName, outerName,
+				)
+			}
+		}
+	}
+
```

Comment:
```suggestion
```

This loops is unnecessary since request/response adding happens in our `evaluateRequest(...)` method in this object. Additionally, history can be inconsistent if an error occurs; right now that triggers the response not to be stored, but we could adjust that if desired.